client/core/trade_simnet_test.go

```diff
@@ -417,37 +416,48 @@ func monitorTradeForTestOrder(ctx context.Context, client *tClient, orderID stri
 			lastProcessedStatus[match.id] = status
 			client.log("NOW =====> %s", status)
 
-			// check for and mine client's swap or redeem
 			var assetToMine *dex.Asset
 			var swapOrRedeem string
+
 			switch {
 			case side == order.Maker && status == order.MakerSwapCast,
 				side == order.Taker && status == order.TakerSwapCast:
-				assetToMine, swapOrRedeem = tracker.wallets.fromAsset, "swap"
+				// Record expected balance changes if we've just sent a swap.
+				// Do NOT mine blocks until counter-party captures status change.
+				recordBalanceChanges(tracker.wallets.fromAsset.ID, true, match.Match.Quantity, match.Match.Rate)
+
+			case side == order.Maker && status == order.TakerSwapCast,
+				side == order.Taker && status == order.MakerSwapCast:
+				// Mine block for counter-party's swap. This enables us to
+				// proceed with the required follow-up action.
+				// Our toAsset == counter-party's fromAsset.
+				assetToMine, swapOrRedeem = tracker.wallets.toAsset, "swap"
+
 			case side == order.Maker && status == order.MakerRedeemed,
 				side == order.Taker && status == order.MatchComplete:
+				recordBalanceChanges(tracker.wallets.toAsset.ID, false, match.Match.Quantity, match.Match.Rate)
+				// Mine blocks for redemption since counter-party does not wait
+				// for redeem tx confirmations before performing follow-up action.
 				assetToMine, swapOrRedeem = tracker.wallets.toAsset, "redeem"
```

Comment:
Relating to the conversation here https://github.com/decred/dcrdex/pull/589#discussion_r466886602.

A previous commit moved swap, redeem tx mining to the counter-party's trade monitoring goroutine. That ensured that party A's swap or redeem is only mined after party B's trade monitoring goroutine has had the chance to handle the status change. In cases where party B should not follow up, we set trade.failErr on party B's trade before mining party A's tx.

As mentioned in the referenced conversation, the exception to the above approach is Taker redeeming Maker's swap without needing to wait for Maker's redeem to get any confs. This sometimes results in Taker redeeming Maker's swap as soon as it is notified of Maker's redeem, and BEFORE Taker's trade monitoring goroutine captures the `MakerRedeemed` status and mines Maker's redeem tx. In one instance where this happened, Maker's balance did not reflect the income from redeem and caused the test to fail with a balance check error.

Redemption txs are now mined immediately by the issuing party since the counter-party may not wait for the redeem tx to get confirmed before performing their counter-redeem and end up not mining the initial redeem.

I think this needs some more fleshing out but this approach makes sense and works well for now. Delay swap mining to enable counter-party check if they should act or not, while mining redeems right away as a wait doesn't really help no one.